com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/CopybookServiceImpl.java

```diff
@@ -16,164 +16,117 @@
-   *     exists.
-   */
-  private void resolveCopybookContent(String requiredCopybookName) {
-    ConfigurationSettingsStorable configurationSettingsStorable =
-        configurationSettingsStorableProvider.get();
-    if (configurationSettingsStorable == null) {
-      publishOnDatabus(requiredCopybookName);
-      return;
+    if (copybookPath.containsKey(requiredCopybookName)) {
+      Path file = copybookPath.get(requiredCopybookName);
+      if (files.fileExists(file)) {
+        sendResponse(requiredCopybookName, files.getContentByPath(file), file);
+        return;
+      } else {
+        copybookPath.remove(requiredCopybookName);
+      }
     }
-
-    Path path = findCopybook(requiredCopybookName, configurationSettingsStorable.getPaths());
-    if (isFileExists(path)) {
-      publishOnDatabus(requiredCopybookName, getContentByPath(path), path);
+    if (DID_OPEN.name().equals(event.getTextDocumentSyncType())) {
+      String cobolFileName = files.getNameFromURI(event.getDocumentUri());
+      clientService
+          .callClient("copybook", cobolFileName, requiredCopybookName)
+          .thenAccept(sendResponse(requiredCopybookName));
     } else {
-      publishOnDatabus(requiredCopybookName);
+      sendResponse(requiredCopybookName, null, null);
     }
   }
 
-  private void publishOnDatabus(String requiredCopybookName, String content, Path path) {
+  private Consumer<List<Object>> sendResponse(String requiredCopybookName) {
+    return result -> dataBus.postData(fetchCopybook(requiredCopybookName, retrieveURI(result)));
+  }
+
+  private void sendResponse(String requiredCopybookName, String content, Path path) {
     dataBus.postData(
         FetchedCopybookEvent.builder()
             .name(requiredCopybookName)
-            .uri(Optional.ofNullable(path).map(Path::toUri).map(URI::toString).orElse(null))
+            .uri(toURI(path))
             .content(content)
             .build());
   }
 
-  private void publishOnDatabus(String requiredCopybookName) {
-    dataBus.postData(FetchedCopybookEvent.builder().name(requiredCopybookName).build());
+  private FetchedCopybookEvent fetchCopybook(String requiredCopybookName, String uri) {
+    if (uri.isEmpty()) {
+      return FetchedCopybookEvent.builder().name(requiredCopybookName).build();
+    }
+    Path file = files.getPathFromURI(uri);
+    if (file == null) {
+      return FetchedCopybookEvent.builder().name(requiredCopybookName).build();
+    }
+    copybookPath.put(requiredCopybookName, file);
+
+    return FetchedCopybookEvent.builder()
+        .name(requiredCopybookName)
+        .uri(toURI(file))
+        .content(files.getContentByPath(file))
+        .build();
+  }
+
+  private String toURI(Path file) {
+    return ofNullable(file).map(Path::toUri).map(URI::toString).orElse(null);
+  }
+
+  private String retrieveURI(List<Object> result) {
+    return result.isEmpty() ? "" : ((JsonPrimitive) result.get(0)).getAsString();
```

Comment:
is result null-safe ?

Reply:
Good catch! So if LSP sends us NULL we will have null here.
It's not clear from LSP spec is it possible or not.

Reply:
It should not send null, but we also shouldn't trust the client

Reply:
I agree with Andrei, plus I think at initialize handshake server receives null